repository-s3/src/main/java/io/aiven/elasticsearch/repositories/s3/S3RepositoryStorageIOProvider.java

```diff
@@ -0,0 +1,270 @@
+
+        private final int partSize;
+
+        private final String bucketName;
+
+        private final CryptoIOProvider cryptoIOProvider;
+
+        private S3StorageIO(final String bucketName, final int partSize, final CryptoIOProvider cryptoIOProvider) {
+            this.bucketName = bucketName;
+            this.partSize = partSize;
+            this.cryptoIOProvider = cryptoIOProvider;
+        }
+
+        @Override
+        public boolean exists(final String blobName) throws IOException {
+            try {
+                return Permissions.doPrivileged(() -> client.doesObjectExist(bucketName, blobName));
+            } catch (final Exception e) {
+                throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists", e);
+            }
+        }
+
+        @Override
+        public InputStream read(final String blobName) throws IOException {
+            return Permissions.doPrivileged(() -> {
+                final S3ObjectInputStream objectContent;
+                try {
+                    objectContent = client.getObject(bucketName, blobName).getObjectContent();
+                } catch (final AmazonClientException e) {
+                    throw new IOException("Couldn't read blob " + blobName, e);
+                }
+                return cryptoIOProvider.decryptAndDecompress(objectContent);
+            });
+        }
+
+        @Override
+        public void write(final String blobName,
+                          final InputStream inputStream,
+                          final long blobSize,
+                          final boolean failIfAlreadyExists) throws IOException {
+            try {
+                Permissions.doPrivileged(() ->
+                        cryptoIOProvider.compressAndEncrypt(
+                                inputStream,
+                                new S3OutputStream(bucketName, blobName, partSize, client)
+                        )
+                );
+            } catch (final AmazonClientException e) {
+                throw new IOException("Couldn't upload blob with name: " + blobName, e);
+            }
+        }
+
+        @Override
+        public Tuple<Integer, Long> deleteDirectories(final String path) throws IOException {
+            final List<Tuple<String, Long>> listOfFiles;
+            try {
+                //we need only prefix here since it is impossible to distinguish between "files" and "directories"
+                //in AWS SDK e.g. paths like:
+                // - /aaaa/bbb/ccc/file
+                // - /aaa/bbb/file
+                // with request which contains delimiter and without returns diff result, and ES API doesn't use
+                // recursive calls to get full list of files by path/sub-path
+                listOfFiles = Permissions.doPrivileged(() ->
+                        list(new ListObjectsV2Request().withBucketName(bucketName).withPrefix(path)))
```

Comment:
Could `listOfObjectsRequest` be used here?

Reply:
Unfortunately not that's why I added comment

Reply:
Gotcha